File: Patterns/Command Pattern/CSharp/Command/CailingFan/CailingFanOnToMediumCommand.cs
namespace Command.CailingFan
{
    public class CailingFanOnToMediumCommand : ICommand
    {
        private CailingFan cailingFan;
        private int preview;

        public CailingFanOnToMediumCommand(CailingFan cailingFan)
        {
            preview = CailingFan.OFF;
            this.cailingFan = cailingFan;
        }

        public void Execute()
        {
            preview = cailingFan.State;
            cailingFan.SetMedium();
        }

        public void Undo()
        {
            switch (preview)
            {
                case CailingFan.HIGH:
                    cailingFan.SetHigh();
                    break;
                case CailingFan.LOW:
                    cailingFan.SetLow();
                    break;
                case CailingFan.OFF:
                    cailingFan.Off();
                    break;
            }
        }
    }
}

File: Patterns/Command Pattern/Cpp/Command/CailingFan/CailingFanOnToMediumCommand.cpp
#include "../ICommand.cpp"
#include "./CailingFan.cpp"

class CailingFanOnToMediumCommand : public ICommand {
private:
    CailingFan* cailingFan;
    int preview;

public:
    CailingFanOnToMediumCommand(CailingFan* cailingFan) : cailingFan{cailingFan}, preview{cailingFan->OFF} {}

    void execute() override {
        preview = cailingFan->getState();
        cailingFan->setMedium();
    }

    void undo() override {
        if (preview == cailingFan->HIGH) cailingFan->setHigh();
        else if (preview == cailingFan->MEDIUM) cailingFan->setMedium();
        else if (preview == cailingFan->OFF) cailingFan->off();
    }

    string getCommandName() override {
        return "CailingFanOnToMediumCommand";
    }
};